magento/magento2
# Fixed duplicated ids to index by mview changelog

### Description (*)
When we have a large number of records in changelog tables  (over 1000) then indexer called by mview action can process this same ids few times. 

`ChangeLogBatchWalker` is changed to use temporary table to collect unique ids to reindex it. List of ids is returns in batches by `Generator`.

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/magento2/issues/30012

### Manual testing scenarios (*)
1. Set indexer in "On Scheduled" mode.
2. Insert to the changelog table related to your indexer over 1000 ids to reindex them. But in on each 1000 records your need have duplicated id. This can be a list of this same id duplicated over 1000x. 
3. Run `indexer_update_all_views` cron job.
4. You will see this same ids used few times by mview action.

In case when your changelog table will be filled by this same id (ex: 1) 10000x then you should see used this id (1) 10x by mview action.

### Related Pull Requests
https://github.com/magento-gl/magento2-infrastructure/pull/22


### Contribution checklist (*)
